### DISCUSSION TITLE
Make `TranslatorExtension` extendable

### DISCUSSION THREAD

**Discussion by daun** [nette/latte#355] [General]
## Explain your intentions

I'd like to extend a custom `TranslatorExtension` from Latte's built-in one. I mainly need to encapsulate the translator function in a class, so that consumers can just do `new CustomTranslator()` without having to pass in a callback themselves. Like below. Unfortunately, this is currently impossible as `TranslatorExtension` is marked as `final`.

```php
class CustomTranslator extends \Latte\TranslatorExtension
{
    public function __construct() {
      parent::__construct([$this, 'translate']);
    }

    public function translate($key, $data = null, $lang = null) {
        return app()->translate($key, $data, $lang);
    }
}
```

## Make a strong case of the merits of this feature

I'm assuming there's no specific reason the class can't be extendable. If one wanted to do this, one would have to re-implement all the useful features inside it, like the `_` tag and the `|translate` filter.

**Comment by dg** [comment]
This is not an optimal use of inheritance. It is better to use either composition or, better yet, factory:

```php
class CustomTranslatorFactory
{
    public static function create() {
      return new Latte\TranslatorExtension([self::class, 'translate']);
    }

    public static function translate($key, $data = null, $lang = null) {
        return app()->translate($key, $data, $lang);
    }
}


$latte->addExtension(CustomTranslatorFactory::create());
```

**Reply by daun** [reply]
Indeed, this is probably how it should be done. Thanks for the example, I'll go with that!
